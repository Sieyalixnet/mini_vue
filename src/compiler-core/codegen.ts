import { NodeTypes } from "./ast"
import { helperMapName, TO_DISPLAY_STRING} from "./runtimeHelpers"
export function generate(ast) {
    let context = createCodeContext()
    const { push } = context

    if(ast.helpers.length>0){
    genFunctionPreamble(ast, context)}

    push('return ')
    const functionName = "render"
    let args = ["_ctx", "_cache"]
    let signature = args.join(', ')

    push(`function ${functionName}(${signature}){`)
    genNode(ast.codegenNode, context)
    push('}')
    return { code: context.code }

}

function genFunctionPreamble(ast, context) {
    const { push } = context
    const VueBeinging = "Vue"
    const helpersAlias = (s) => { return `${helperMapName[s]}: _${helperMapName[s]}` }
    const helpers = ast.helpers.map(helpersAlias).join(", ")
    push(`const {${helpers}} = Vue`)
    push("\n")

}

function createCodeContext() {
    const context = {
        code: "",
        push(content) { context.code += content }

    }

    return context
}

function genNode(node, context) {
    switch (node.type) {
        case NodeTypes.TEXT:
            genText(node, context)
            break;
        case NodeTypes.INTERPOLATION:
            genInterpolation(node, context)       
            break;
        case NodeTypes.SIMPLE_EXPRESSION:
            genExpression(node, context)
            break;   
        default:
            break;
    }


}

function genText(node, context){
    const { push } = context
    push(`return '${node.content}'`)  

}

function genInterpolation(node, context){
    const { push } = context
    push(`return _${helperMapName[TO_DISPLAY_STRING]}(`)
    genNode(node.content,context)
    push(`)`)  

}

function genExpression(node, context){
    const { push } = context
    push(`${node.content}`)  

}