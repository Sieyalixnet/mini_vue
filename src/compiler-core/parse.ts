import { NodeTypes } from "./ast"

const enum TagTypes {
    START,
    END
}

export function baseParse(content) {

    let context = createParsercontext(content)
    return createRoot(parseChildren(context, []))

}

function parseChildren(context, ancestors) {

    let nodes: any = []
    let node
    while (!isEnd(context, ancestors)) {
        node = undefined
        if (context.source.startsWith("{{")) {
            console.log("Interpo")
            node = parseInterpolation(context)
        } else if (context.source.startsWith("<")) {
            console.log("Element")
            if (/[a-z]/i.test(context.source[1])) {
                node = parseElement(context, ancestors)
            }
        }

        if (!node) {
            console.log("text")
            node = parseText(context)
        }
        nodes.push(node)
    }


    return nodes;

}

function isEnd(context, ancestors) {
    if (ancestors.length > 0) {
        for (let i = ancestors.length - 1; i >= 0; i--) {
            if (startsWithEndTag(context, ancestors[i].tag)) { return true }

        }
    }

    return !context.source //有值时为false; 当context.source为""时, 本身为false,加上!会变为true

}

function parseText(context) {
    let endIndex = context.source.length
    let endMarks = ["{{", "<"]
    //注: 不同点. 我这里用的是高阶函数. 也可以用回for方法
    let endMark = endMarks.map((s) => { return context.source.indexOf(s) }).sort((s1, s2) => { return s1 - s2 }).find((s) => { return s != -1 })
    //一定要把最近那个标号放前面来,然后再find
    if (endMark) {
        endIndex = endMark
    }

    const content = parseTextData(context, endIndex)
    return {
        type: NodeTypes.TEXT,
        content
    }

}

function parseTextData(context, length) {
    const content = context.source.slice(0, length)
    advanceBy(context, content.length)
    return content
}

function parseElement(context, ancestors) {
    const element: any = parseTag(context, TagTypes.START)
    ancestors.push(element)//WRONG IN 20220524 放错地方. 是在拿到tag之后马上就需要给ancestors加进去, 在parseTag之前就弹出.

    element.children = parseChildren(context, ancestors)
    ancestors.pop()
    if (startsWithEndTag(context, element.tag)) { parseTag(context, TagTypes.END) } else {
        throw (`${element.tag} lacks of end tag`)
    }

    return element
}

function startsWithEndTag(context, tag) {

    return context.source.startsWith("</") && context.source.slice(2, 2 + tag.length) === tag

}

function parseTag(context, TagTypes) {

    const reg = /^<\/?([a-z]*)/i
    const matchResult: any = reg.exec(context.source)
    let tag = matchResult[1]
    advanceBy(context, matchResult[0].length)
    advanceBy(context, 1)
    if (TagTypes = TagTypes.END) { return; }
    return {
        type: NodeTypes.ELEMENT,
        tag
    }


}

function parseInterpolation(context) {

    const openDelemiter = "{{"
    const closeDelemiter = "}}"

    let closeIndex = context.source.indexOf(closeDelemiter, closeDelemiter.length)//找-找到闭合的index
    advanceBy(context, openDelemiter.length)//推-掉本次处理的前置符号
    let rawContent = parseTextData(context, closeIndex - closeDelemiter.length)

    //拿-拿到所需要的部分
    let content = rawContent.trim()
    advanceBy(context, closeDelemiter.length)//推-推至本次处理的后置部分
    return {
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: content
        }

    }

}

function advanceBy(context, advanceIndexStep) {
    context.source = context.source.slice(advanceIndexStep)
}

function createRoot(children) {

    return { children,
    type:NodeTypes.ROOT
    }

}


function createParsercontext(content) {

    return { source: content }

}