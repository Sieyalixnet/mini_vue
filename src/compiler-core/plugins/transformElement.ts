import { createVNodeCall, NodeTypes } from "../ast"

export function transformElement(node,context){//实际就是把content里的content拿出来, 然后在前面加一层ctx就完成了. 
    if(node.type===NodeTypes.ELEMENT){


        return ()=>{
            
        const VNodeTag = `'${node.tag}'`


        const VNodeProps = node.props
        
        const children = node.children
        const VNodeChildren = children[0]

        const VNodeElement= createVNodeCall(context, VNodeTag,VNodeProps,VNodeChildren)
        node.codegenNode= VNodeElement}
}
}
