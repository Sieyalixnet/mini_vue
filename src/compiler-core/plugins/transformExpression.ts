import { NodeTypes } from "../ast"



export function transformInterpolation(node){
    if(node.type===NodeTypes.INTERPOLATION){
    node.content=processInterpolation(node.content)
}
}

function processInterpolation(node){
    node.content =  `_ctx.${node.content}`
    return node

}