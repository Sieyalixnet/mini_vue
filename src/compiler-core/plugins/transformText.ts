import { NodeTypes } from "../ast"
import { isText } from "../utils"


export function transformText(node, context) {


    return () => {
        const { children, content, tag, type } = node
        if(children){
        for (let i = 0; i < children.length; i++) {
            let child = children[i]
            let currentContainer
            if (isText(child)) {
                for (let j = i + 1; j < children.length; j++) {
                    let next = children[j]
                    if (isText(next)) {
                        if (!currentContainer) {//必须放j的判定之后， 因为只有连续2个子节点是符合isText才需要这个container。在i中写其实也还是要判断，多此一举，所以直接在j的循环中最合理。
                            currentContainer = children[i] = {//下面的else引用类型的特点， 也就是currentContainer被改了之后, children[i]是不会变的
                                type: NodeTypes.COMPOUND_EXPRESSION,
                                children: [child]
                            }
                        }

                        currentContainer.children.push(" + ")
                        currentContainer.children.push(next)

                        children.splice(j, 1)
                        j--
                    } else {
                        currentContainer = undefined
                        break;
                    }
                }
            }
        }
    }}
}