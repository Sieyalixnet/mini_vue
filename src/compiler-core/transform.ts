import { NodeTypes } from "./ast"
import { TO_DISPLAY_STRING } from "./runtimeHelpers"


export function transform(root, options={}) {
    const context = createTransformContext(root, options)
    traverseNode(root, context)
    createRootCodegen(root)
    root.helpers = [...context.helpers.keys()]

}
function createRootCodegen(root){
    root.codegenNode =root.children[0]

}

function createTransformContext(root: any, options: any) {
    return {
        root,
        nodeTransfroms: options.nodeTransfroms || [],
        helpers:new Map(),
        helper(key){
            this.helpers.set(key,1)
        }
    }

}

function traverseNode(node, context) {
    let children = node.children
    const { nodeTransfroms } = context
    for (let i = 0; i < nodeTransfroms.length; i++) {
        nodeTransfroms[i](node)
    }
    switch (node.type) {
        case NodeTypes.INTERPOLATION:
            context.helper(TO_DISPLAY_STRING)
            break;
        case NodeTypes.ROOT:
        case NodeTypes.ELEMENT:
            traverseChildren(children, context)
        default:
            break;
    }
    

}

function traverseChildren(children, context) {
//事实上它只是遍历一下, 找到符合的点就会改. 没符合就遍历过一下而已
        for (let i = 0; i < children.length; i++) {
            const node = children[i]
            traverseNode(node, context)
        }
}