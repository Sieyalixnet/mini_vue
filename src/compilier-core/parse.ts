import { NodeTypes } from "./ast"

const enum TagTypes {
    START,
    END
}

export function baseParse(content) {

    let context = createParsercontext(content)
    return createRoot(parseChildren(context))

}

function parseChildren(context) {

    let nodes: any = []
    let node
    if (context.source.startsWith("{{")) {
        node = parseInterpolation(context)
    } else if (context.source.startsWith("<")) {
        if (/[a-z]/i.test(context.source[1])) {
            node = parseElement(context)
        }

    }
    nodes.push(node)

    return nodes;

}

function parseElement(context) {
    const element = parseTag(context, TagTypes.START)
    parseTag(context, TagTypes.END)
    return element
}

function parseTag(context, TagTypes) {

    const reg = /^<\/?([a-z]*)/i
    const matchResult: any = reg.exec(context.source)
    let tag = matchResult[1]
    advanceBy(context, matchResult[0].length)
    advanceBy(context, 1)
    if (TagTypes = TagTypes.END) { return; }
    return {
        type: NodeTypes.ELEMENT,
        tag
    }


}

function parseInterpolation(context) {

    const openDelemiter = "{{"
    const closeDelemiter = "}}"

    let closeIndex = context.source.indexOf(closeDelemiter, closeDelemiter.length)//找-找到闭合的index
    advanceBy(context, openDelemiter.length)//推-掉本次处理的前置符号
    let rawContent = context.source.slice(0, closeIndex - closeDelemiter.length)//拿-拿到所需要的部分
    let content = rawContent.trim()
    advanceBy(context, closeIndex + closeDelemiter.length)//推-推至本次处理的后置部分
    return {
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: content
        }

    }

}

function advanceBy(context, advanceIndexStep) {
    context.source = context.source.slice(advanceIndexStep)
}

function createRoot(children) {

    return { children }

}


function createParsercontext(content) {

    return { source: content }

}