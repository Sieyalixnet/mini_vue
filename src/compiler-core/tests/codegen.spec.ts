import { generate } from "../codegen"
import { baseParse } from "../parse"
import { transformInterpolation } from "../plugins/transformExpression"
import { transform } from "../transform"

describe("code generator", () => {

    test("string", () => {
        const ast = baseParse('hi')
        transform(ast)
        const {code} = generate(ast)
        
        expect(code).toMatchSnapshot()


    })

    test("interpolation", () => {
        const ast = baseParse('{{message}}')
        transform(ast,{nodeTransfroms:[transformInterpolation]})
        const {code} = generate(ast)
        
        expect(code).toMatchSnapshot()


    })
})