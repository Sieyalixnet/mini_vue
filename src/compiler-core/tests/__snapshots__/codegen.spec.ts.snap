// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code generator interpolation 1`] = `
"const {toDisplayString: _toDisplayString} = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`;

exports[`code generator string 1`] = `"return function render(_ctx, _cache){return 'hi'}"`;
